Native pipe

vors opened the pull request:
Fix #559 
Improve pipeline for native commands
- Start native process in Prepare() instead of Complete()
  in NativeCommandProcessor.
- Remove unneeded input, output, error threads. Replaced by
  simpler primitives.

Motivation example:

`ping | %{powershell-code} | grep` now works in real-time.
See the [test](https://github.com/PowerShell/PowerShell/pull/2450/commits/e34a81a30fda95685084b09d4a0d5f60006e4d49#diff-e2e61256a6481fbfcebf5757b220793aR35)

TODO before merging:

- [x] replace spin-lock by something like `notify/wait`
- [x] resurrect and test minishell with custom host serialization
- [x] Changelog entry
- [x] Fix feature tests for https://github.com/PowerShell/PowerShell/blob/ae7b857c235203d1fd3708fe05c076e5a376859a/test/powershell/Host/ConsoleHost.Tests.ps1#L57
- [x] Investigate, is this test flaky or not https://ci.appveyor.com/project/PowerShell/powershell/build/6.0.0-alpha.12-1287/tests

Commit:
Add a minishell clixml tests

Commit:
Improve pipeline for native commands

- Start native process in Prepare() instead of Complete()
  in NativeCommandProcessor.
- Remove unneeded input, output, error threads. Replaced by
  simpler primitives.

msftclas commented:
Hi __@vors__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Sergei Vorobev). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

Commit:
Add error stream and information stream tests

for native pipes

Commit:
Make native pipeline consume available output in the process block

Commit:
Add motivation example for the native pipeline changes

Commit:
Fix flushing mechanics for input writer in native pipe

Commit:
Use StreamWriter.Dispose() instead of Close()

daxian-dbw commented:
@vors could you please resolve the Travis CI failures?

vors commented:
@daxian-dbw working on it.
There are two parts: PSDSC.psm1 need to be updated: it uses `sort` instead of `sort-object` in one place and worked by accident. The second part is about handing native executable failures because it still should not produce broken pipe error.

Commit:
Add a test for broken linux pipe in native pipeline

Commit:
Add a code to handle finished process correctly on Unix

vors commented:
🎉 @daxian-dbw ready for review

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1104,7 +1364,7 @@ private bool IsDownstreamOutDefault(Pipe downstreamPipe)
         private void CalculateIORedirection(out bool redirectOutput, out bool redirectError, out bool redirectInput)
         {
-            redirectInput = true;
+            redirectInput = this.Command.MyInvocation.PipelineLength > 0;
```
The native command may be the first command in the pipeline, and it doesn't need input in that case.

vors replied:
Good catch, it should be 'PipelinePosition'

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1352,31 +1609,46 @@ internal ProcessInputWriter(InternalCommand command)
             Dbg.Assert(command != null, "Caller should validate the parameter");
             _command = command;
+
+            _pipeline = ScriptBlock.Create("Out-String -Stream").GetSteppablePipeline();
```
Will `-Stream` be a breaking change?

vors replied:
No

daxian-dbw replied:
why? `Out-String` returns a single string, while `Out-String -Stream` returns an string array.

vors replied:
Hmmm. Ok, so I would need to double check this. Here is my original thinking:

Before we had all data upfront, so we can use just `Out-String`. Now we get new data sporadically and should still output something, even when data is incomplete. `-Stream` seemed like a right switch.

After your comments, I revisited this thinking and it could be indeed incorrect. I think data would still be available, there would be no aggregation. For example, lets take a look at

```
$h = @{1 = 2}; $h; sleep 1; $h | Out-String | %{$_}
```

When I run it, I get the first `$h` immediately.

Also `-Stream` adds overhead, because there are more objects. So I think you are correct. Thank you for spotting the problem!

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1352,31 +1609,46 @@ internal ProcessInputWriter(InternalCommand command)
+            foreach (var item in formattedObjects)
             {
-                return _inputList.Count;
+                string line = PSObject.ToStringParser(_command.Context, item);
```
It looks you are not differentiating `Text` input from `Xml` input, but you are differentiating them when processing the output of native command. Is this a oversight?

vors replied:
Possible, but I found that serizliation is handled by the code in `src\Microsoft.PowerShell.ConsoleHost\host\msh\Serialization.cs`. I think serialization path here was not used at all. I'd love to see proven wrong, if you can find a case, where it was used.

daxian-dbw replied:
`XmlInput` has been brought back in the new iteration.
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1352,31 +1609,46 @@ internal ProcessInputWriter(InternalCommand command)
+
+            _pipeline = ScriptBlock.Create("Out-String -Stream").GetSteppablePipeline();
+            _pipeline.Begin(true);
         }
```
With your change to `ProcessInputWriter`, we won't be able to use the same InputWriter type to handle byte stream transfers between native commands -- we will need a thread to read byte output from upstream native command and write to the downstream command.
My opinion is to keep the "use-separate-thread" module of `ProcessInputWriter`, but again use a `BlockingCollection` between `Add` and the thread that does the actual write operation. It's essentially a consumer/provider problem.

vors replied:
The existing approach was spinning a thread and managing its lifecycle. It's pretty wasteful and unnecessary complicated. One of the my goal is to avoid a "medium" between two native processes, while implementing #1908. It would be a simple redirection without marshaling. So this would be a completely different code path. If we will try to reuse existing approach with `ProcessInputWriter` and enhance it for bytes, powershell would never be able to get closer to `bash` in `native | native` case by it's performance. We need to strip away the overhead of marshaling objects and just do the stream redirection.

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                        {
+                            bool isFirstOutput = true;
+                            bool isXmlCliOutput = false;
+                            _nativeProcess.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
```
What you are describing is #1908 and it's outside of the scope of this PR. I'd like to solve it with creating already piped processes and avoid expensive manual marshaling from stdout of one to stdin of another.

daxian-dbw replied:
Yes, I know #1908 is not covered by this PR. My point is that we would still need a thread worker to transfer byte streams between native commands, so probably we should keep the `ProcessOutputReader` implementation there for now just in case that we need to reuse some of that code.
And for the same reason, using `OutputDataReceived` and `ErrorDataReceived` to collect process output/error may not be appropriate since we probably have to change it again using a thread worker when addressing #1908

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                    {
+                        // TODO: Currently the main pipeline Thread is just spinning.
+                        // It would be much better to park it until new input is available.
+                        ConsumeAvailableNativeProcessOutput();
```
`ConcurrentQueue` has the same blocking interface: `Dequeue` (instead of `TryDequeue`), but it's not the point of this TODO note. We should not block, because process may just stop at any moment without producing output. The right way to do it would be to use wait/notify where notify will be called the code that adds to collection and by a callback about process finish.

daxian-dbw replied:
`BlockingCollection` provides `CompleteAdding` (https://msdn.microsoft.com/en-us/library/dd287086(v=vs.110).aspx), so when the process is gone, we can call this method to notify the consumer that adding is done. In that case, `Take` from the collection is not blocking when the collection is empty. Take your current design as an example, when the output stream of the underlying native command is closed, `e.Data` would be `null`, and when seeing that, you can call `CompleteAdding` on the collection, and in `Complete`, you can take `Take` from the collection -- it will block when the collection is empty unless the `CompleteAdding` is called. So you don't need to handle the wait/notify yourself.

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
-                exceptionToRethrow = e;
+                _exceptionToRethrow = e;
             } // try
             catch (PipelineStoppedException)
```
Maybe. I though that `CommandProcessorBase.CheckForSevereException` is valuable check, but it may be not.

vors replied:
make sense :)

daxian-dbw replied:
Ah, good point. I didn't notice the `catch all` down there.

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -140,6 +142,8 @@ internal ProcessOutputObject(object data, MinishellStream stream)
     /// </summary>
     internal class NativeCommandProcessor : CommandProcessorBase
     {
+        private static string XmlCliTag = "#< CLIXML";
```
done

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -331,17 +338,39 @@ internal override void ProcessRecord()
+        /// Indicate whether we need to consider redirecting the output/error of the current native command.
+        /// Usually a windows program which is the last command in a pipeline can be executed as 'background' -- we don't need to capture its output/error streams.
+        /// </summary>        
+        private bool _background;
```
sure

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -331,17 +338,39 @@ internal override void ProcessRecord()
+        /// </summary>        
+        private bool _background;
+
+        private ProcessStartInfo _startInfo;
```
great idea!

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -331,17 +338,39 @@ internal override void ProcessRecord()
+
+        private bool _scrapeHostOutput;
+
+        private bool _redirectInput;
```
Good catch!

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -331,17 +338,39 @@ internal override void ProcessRecord()
+        /// If a problem occurred in running the program, this exception will
+        /// be set and should be rethrown at the end of the try/catch block...
+        /// </summary> 
+        private Exception _exceptionToRethrow = null;
```
sure

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+        /// </returns>
+        private bool ConsumeAvailableNativeProcessOutput()
+        {
+            bool isNewData = false;
```
I removed this var

daxian-dbw replied:
#Closed

vors commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -921,84 +1191,74 @@ private void CleanUp()
 
-            object value = _outputReader.Read();
-            while (value != AutomationNull.Value)
+            
```
done

Commit:
Clean-up unnesesary fields from NativeCommandProcessor

vors commented:
I will update PR with the `notify/wait` code

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                _exceptionToRethrow = e;
+
+            } // try
+            catch (PipelineStoppedException)
```
By design #Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+        /// <returns>
+        /// True if there was any new input available, otherwise false.
+        /// </returns>
+        private bool ConsumeAvailableNativeProcessOutput()
```
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -921,84 +1191,74 @@ private void CleanUp()
+            if (this.Command.Context.CurrentPipelineStopping)
+            {
+                this.StopProcessing();
+                return;
```
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1405,27 +1658,22 @@ internal int Count
-            }
-            _inputThread = new Thread(new ThreadStart(this.WriterThreadProc));
-            _inputThread.Start();
+            _inputFormat = inputFormat;
```
`_inputFormat` was used for processing `Text` or `Xml` input. If it's proven that we can ignore `Xml` input, then we need to clean up `ProcessInputWriter` to remove this member, including the parameter of `Start` method.

vors replied:
I found the case, when we use this code:

```powershell
1,2,3 | powershell -noprofile -command { $input }
```

Resurrecting it now

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1352,31 +1595,46 @@ internal ProcessInputWriter(InternalCommand command)
+                catch (IOException)
+                {
+                    // we are assuming that process is already finished
+                    // we should just stop processing at this point
```
Minor: may be we can avoid this exception by using the overload `StreamWriter(Stream stream, Encoding encoding, int bufferSize, bool leaveOpen);`. The writer will not try to dispose the stream when `leaveOpen = true`. Let the process object handle closing the stream.

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1439,574 +1687,35 @@ internal void Stop()
+            // we allow call Done() multiply times.
+            // For example one time from Process() code path, 
+            // when we detect that process already finished
+            // and once from End() code path.
```
`Process()` and `End()` is kinda confusing at first glance. Can you please change `Process()` to `ProcessRecord()` and `End()` to `Complete()`?
And please add one more comment explains that "even though `Done` could be called multiple times, the calls are on the same thread, so there is no race condition here".

vors replied:
Done

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1439,574 +1687,35 @@ internal void Stop()
-                    WriteXmlInput();
+                    // on unix, if process is already finished attempt to dispose it will
+                    // lead to "Broken pipe" exception.
+                    // we are ignoring it here
```
Minor: may be we can avoid this exception by using the overload StreamWriter(Stream stream, Encoding encoding, int bufferSize, bool leaveOpen);. The writer will not try to dispose the stream when leaveOpen = true. Let the process object handle closing the stream.

vors replied:
I dig into that and as far as I see, that will hang the process: there is no way to tell it that all input was present and close the stream, if process expects input.

daxian-dbw replied:
Good point. #Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,346 @@ internal override void Complete()
+                                    if (isFirstOutput)
+                                    {
+                                        isFirstOutput = false;
+                                        if (e.Data == XmlCliTag)
```
It's recommended to a culture overload `Equals` method to compare strings.

vors replied:
Done

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,346 @@ internal override void Complete()
+                                if (e.Data != null)
                                 {
-                                    lineContents.Append(bufferContents[row, column].Character);
+                                    if (e.Data == XmlCliTag)
```
Same here.

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1352,31 +1595,46 @@ internal ProcessInputWriter(InternalCommand command)
-            get
+            Array formattedObjects = _pipeline.Process(input);
+            foreach (var item in formattedObjects)
             {
```
Shall we add `if (_stopping) { return }` in this loop?

daxian-dbw replied:
#Closed

daxian-dbw commented on test/powershell/Language/Scripting/NativeExecution/NativeMinishell.Tests.ps1:
```diff
@@ -0,0 +1,50 @@
+# Minishell is a powershell concept.
+# It's primare use-case is when somebody executes a scriptblock in the new powershell process.
```
You mean `Its primary`?

vors replied:
Done

daxian-dbw replied:
You missed the typo `primare`.

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,346 @@ internal override void Complete()
+                        {
+                            bool isFirstOutput = true;
+                            bool isXmlCliOutput = false;
+                            _nativeProcess.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
```
Can we encapsulate the event handlers and related variables/methods/objects into a class? I'm thinking about something like the following:
```CSharp
...
var handler = new ProcessOutputHandler(true, true);
_nativeProcess.OutputDataReceived += new DataReceivedEventHandler(handler.OutputHandler);
_nativeProcess.ErrorDataReceived += new DataReceivedEventHandler(handler.ErrorHandler);
...

public class ProcessOutputHandler
{
    public ProcessOutputHandler (bool redirectOutput, bool redirectError)
    {
        Debug.Assert(redirectOutput || redirectError, "Caller should redirect at least one stream");
        _blockingQueue = new BlockingCollection<Tuple<string, int>>(new ConcurrentQueue<Tuple<string, int>>());

        if (redirectOutput) { _refCount ++; }
        if (redirectError) { _refCount ++; }
    }

    private int _refCount = 0;
    private bool _isFirstOutput = true, _isFirstError = true;
    private bool _isOutputXml = false, _isErrorXml = false;
    private BlockingCollection<Tuple<string, int>> _blockingQueue;

    public BlockingCollection<Tuple<string, int>> BlockingQueue
    {
        get { return _blockingQueue; }
    }

    private List<ProcessOutputObject> DeserializeCliXmlObject(string xml, bool isOutput)
    {
        // ...
    }

    public void OutputHandler(object sendingProcess, DataReceivedEventArgs outLine)
    {
        if (outLine.Data == null)
        {
            // StandardOutput is closed
            if (Interlocked.Decrement(ref _refCount) == 0)
            {
                _blockingQueue.CompleteAdding();
            }
        }
        else 
        {
            // Process output and add to _blockingQueue ...
        }
    }

    public void ErrorHandler(object sendingProcess, DataReceivedEventArgs errorLine)
    {
        if (errorLine.Data == null)
        {
            // ErrorOutput is closed
            if (Interlocked.Decrement(ref _refCount) == 0)
            {
                _blockingQueue.CompleteAdding();
            }
        }
        else
        {
            // Process error and add to _blockingQueue ...
        }
    }
}
```

vors replied:
Good idea, done

daxian-dbw replied:
#Closed

Commit:
Replace spin-lock by a proper BlockingCollection in NativeCommandProcessor

vors commented:
@daxian-dbw I removed spin-lock and employed `BlockedCollection` as you suggested.

Commit:
Add native pipelines entry to changelog.md

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+            }
+        }
+
+        private class ProcessOutputHandler
```
Can you please move this class outside of `NativeCommandProcessor`? My arguments:
1. Nested class usually worsens readability of the code.
2. It's essentially the counterpart of `ProcessInputWriter`, and thus I think it should be a non-nested internal class similar as `ProcessInputWriter`

vors replied:
sure

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+                }
+            }
+
+            private void OutputHandler(object sender, DataReceivedEventArgs e)
```
Minor comment: I always encourage people to use descriptive/meaningful variable names. For example, the parameter `e` here and in `ErrorHandler` below could be a bit more descriptive, like `outputReceived` and `errorReceived`.

vors replied:
Good point

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+            private bool _isFirstError;
+            private bool _isXmlCliOutput;
+            private bool _isXmlCliError;
+            private string _path;
```
Minor comment: again about variable name. `_processFileName` would be better I think 😄

vors replied:
Sure

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
-                        //Wait for input writer to finish.
-                        _inputWriter.Done();
+                // collection is empty or exception been raised
+                return null;
```
This `if (blocking)` block is basically the same as following:
```CSharp
if (blocking)
{
    // if adding was completed and the queue is empty, there is no need to do a blocking Take(),
    if (!_nativeProcessOutputQueue.IsCompleted)
    {
         try
         {
             return _nativeProcessOutputQueue.Take();
         }
         catch (InvalidOperationException)
         {
             // It's a normal situation: another thread can mark collection as CompleteAdding
             // in a concurrent way and we will rise an exception in Take().
             // Although it's a normal situation it's not the most common path
             // and will be executed only on the race condtion case.
         }
    }
    // collection is empty and adding completed, or exception been raised
    return null;
}
```

vors replied:
Oh, cool

vors replied:
Had a conversation with @lzybkr about perf here.
Leaving a common path outside of `try {}` will have a slight positive upside. I think it's worth to leave it as is, I will add a comment with explanation.

daxian-dbw replied:
> Bear in mind that this has nothing to do with try/catch blocks: **you only incur the cost when the actual exception is thrown.** You can use as many try/catch blocks as you want. Using exceptions gratuitously is where you lose performance.

This is from a MSDN article [Performance Tips and Tricks in .NET Applications](https://msdn.microsoft.com/en-us/library/ms973839.aspx?ppud=4). It indicates that try/catch block itself doesn't incur any performance overhead, as long as no exception is thrown.

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+                    //Wait for the process to exit and consume available output
+                    while (!_nativeProcess.HasExited)
+                    {
+                        ConsumeAvailableNativeProcessOutput(blocking: true);
```
This shouldn't be within `while (!_nativeProcess.HasExited)`. It's possible that the process quickly writes a lot output and exits, but the output has been captured in our `_nativeProcessOutputQueue`, and we can still read all those output and process them even if the process is gone. Otherwise, we might miss some some output/error because we stop reading the queue when the process is gone.

vors replied:
Good catch! It's a left-over from the previous spin-lock approach.

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +541,431 @@ internal override void Complete()
-                            StringBuilder bufferText = new StringBuilder();
+                    // read all the available output one more time
+                    _nativeProcess.WaitForExit();
+                    ConsumeAvailableNativeProcessOutput(blocking: true);
```
This `ConsumeAvailableNativeProcessOutput(blocking: true);` won't be necessary. One blocking reading of the queue is guaranteed to read all output/error out of it.

daxian-dbw replied:
#Closed

daxian-dbw commented:
Could you please also update your TODO list in the PR description?

Commit:
Re-shuffle minishell tests

Commit:
Prettify NativeCommmandProcessor.cs

- Move ProcessOutputHandler outside of NativeCommmandProcessor
- Remove redundent ConsumeAvailableNativeProcessOutput() blocking call

Commit:
Resurect Xml serialization logic inside ProcessInputWriter

Commit:
Fix bug in native pipe, where we loose End() output

vors commented:
@daxian-dbw Ready for the final review iteration!

Commit:
Fix recursive Done() problem in steppable pipeline finalization

Commit:
Trying to make 'native command lifecycle' test unflaky

Also move it to the Feature category

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
-        /// Use Done to ensure that all reading threads are finished.
-        /// </summary>
-        internal void Stop()
+        private void AddTextInput(object input)
```
This method is called from only one place and probably will be inlined by the compiler anyway. How about removing this method, and use `AddTextInputFromFormattedArray(_pipeline.Process(input));` directly in `Add(object input)`?

vors replied:
I think it's much more readable with clear `AddTextInput` and `AddXmlInput` separation.
I trust the compiler to do the right thing :)

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
+                // if the downstream process already exited.
+                // to Prevent it, we first finalize the pipeline and set it to null,
+                // then calling the result processing for the last time
+                AddTextInputFromFormattedArray(finalResults);
```
speaking of try/catch perf penalty, when `Done()` is called from `AddTextInputFromFormattedArray`, this will raise another exception for sure, which should be avoided.
Two questions:
1. Is calling `_pipeline.End()` really absolutely necessary? It seems `_pipeline.End()` for `Out-String -stream` only write out some newlines. Is there a regression without those new lines?
2. If it's really necessary, I suggest having `Done` renamed to be `Dispose`:
```
internal Dispose()
{
    if (_pipeline != null)
    {
        _pipeline.Dispose();
        _pipeline = null;
    }

    // ... Same as before in below.
}
```
Then have another method named `Done()`
```
internal Done()
{
    if (_pipeline != null)
    {
        AddTextInputFromFormattedArray(_pipeline.End());
    }
    Dispose()
}
```
So in `AddTextInput` and `AddXmlInput`, it calls `Dispose()` when the IO exception is caught, while in `Complete`, it calls `Done()`.

Another alternative is to add a parameter to the current `Done()` method, so that if it's called from `AddTextInput` or `AddXmlInput`, don't call `_pipeline.End()` and `AddTextInputFromFormattedArray`

vors replied:
>  Is calling _pipeline.End() really absolutely necessary? It seems _pipeline.End() for Out-String -stream only write out some newlines. Is there a regression without those new lines?

Yes, that's what I was thinking! It turns out that there are cases, where `End()` has a whole bunch of useful lines. One of the tests catch it.

vors replied:
Good suggestions. 👍 👍 
I also don't like how cumbersome this code became.
Let me re-think it a little bit and I'll probably do something similar to your suggestion.

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
-                    _writer.Write(dataObject);
+                    // on unix, if process is already finished attempt to dispose it will
+                    // lead to "Broken pipe" exception.
+                    // we are ignoring it here
```
Again, speaking of exception perf penalty, how often this will happen on unix platforms? If it's often enough, maybe checking `process.HasExited` before disposing (only for unix) would be preferable?

vors replied:
It's once per process, so not a big of a deal.
For our original exception perf discussion, it was once for every line of text we pipe between processes, which could be huge.

daxian-dbw commented:
In `NativeMinishell.Tests.ps1`:
> Its primare use-case

You missed the typo **primare**.

Commit:
Separate Done() and Dispose() in ProcessInputWriter NativeCommandProcessor.cs

vors commented:
Ready for review!

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,448 @@ private bool IsMiniShell()
-                // nothing to do
-            }
-            else if (line.Equals("#< CLIXML", StringComparison.Ordinal) == false)
+            // we allow call Done() multiply times.
```
`Done()` should be `Dispose()` here.

vors replied:
good catch

daxian-dbw replied:
#Closed

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,448 @@ private bool IsMiniShell()
-            _processOutput = new ObjectStream(128);
+            if (_stopping || _streamWriter == null)
+            {
+                // if _streamWriter is already null, then we already called Done()
```
`Done()` should be `Dispose()` here.

vors replied:
Good catch

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1344,676 +1431,448 @@ private bool IsMiniShell()
-                    }
-                    AddObjectToWriter(obj, stream);
+                    _xmlSerializer.Done();
+                    Dispose();
```
This `Dispose()` is unnecessary as it's called in the end of this method.

vors replied:
oh, good catch! I didn't remove it from here

daxian-dbw commented:
I don't see your updates on the duplicate `_nativeProcessOutputQueue.Take()`. I still think it's unnecessary duplication of code.

Commit:
Remove over-optimization in NativeCommandProcessor.DequeueProcessOutput

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -528,90 +539,179 @@ internal override void Complete()
+        {
+            if (blocking)
+            {
+                if (!_nativeProcessOutputQueue.IsCompleted)
```
Can we please keep the old comment `// if adding was completed and the queue is empty, there is no need to do a blocking Take()`? It will help reader to understand why we check `IsCompleted` which is a great optimization to avoid unnecessary exception from `Take()`.

daxian-dbw replied:
#Closed

daxian-dbw reviewed (approved):
LGTM. Thank you for getting this work done!

iSazonov commented:
Great! :star2:
